Update tokio-core uses to tokio-0.1

Marwes opened the pull request:
`reqwest` has received an update to drop its `tokio-core` dependency so it is now possible to move to using `tokio` here as well.

`hyperx` were pulled it to paper over `hyper`'s removal of typed headers but were otherwise avoided.

Since there are quite a few non-Send futures I had to use `current_thread::Runtime` instead of the standard multi-threaded runtime.

luser commented:
I'm probably missing some bit of information, but what's the motivation behind these changes?

Marwes commented:
Just a follow up of https://github.com/mozilla/sccache/pull/286. `tokio-core` is deprecated with `tokio` as the replacement.

Marwes commented on src/server.rs:
```diff
@@ -266,14 +264,44 @@ impl<C: CommandCreatorSync> SccacheServer<C> {
-        let server = listener.incoming().for_each(move |(socket, _addr)| {
+        let server = listener.incoming().for_each(move |socket| {
             trace!("incoming connection");
-            SccacheProto.bind_server(&handle, socket, service.clone());
```
Forgot I need to cleanup this part. `tokio-proto` is unmaintained so I had to string together a minimal version of it instead but it still uses some of tokio-proto's types which makes it a bit of a mess.

https://github.com/tokio-rs/tokio-proto.

Marwes commented:
Seeing a deprecation warning about `length_delimited` but fixing is blocked on https://github.com/antoyo/tokio-serde-bincode/issues/1

```
warning: use of deprecated item 'tokio_io::codec::length_delimited::Framed': Moved to tokio-codec                                                                                                           
  --> src/server.rs:57:5                                                                                                                                                                                    
   |                                                                                                                                                                                                        
57 | use tokio_io::codec::length_delimited::Framed;                                                                                                                                                         
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                          
   |                                                                                                                                                                                                        
   = note: #[warn(deprecated)] on by default                                                                                                                                                                
                                                                                                                                                                                                            
warning: use of deprecated item 'tokio_io::codec::length_delimited::Framed': Moved to tokio-codec                                                                                                           
   --> src/server.rs:969:37                                                                                                                                                                                 
    |                                                                                                                                                                                                       
969 |     inner: WriteBincode<ReadBincode<Framed<I>, Request>, Response>,                                                                                                                                   
    |                                     ^^^^^^^^^                                                                                                                                                         
                                                                                                                                                                                                            
warning: use of deprecated item 'tokio_io::codec::length_delimited::Framed': Moved to tokio-codec                                                                                                           
   --> src/server.rs:475:55                                                                                                                                                                                 
    |                                                                                                                                                                                                       
475 |             inner: WriteBincode::new(ReadBincode::new(Framed::new(socket))),          
    |                                                       ^^^^^^^^^^^
                                                                                       
warning: use of deprecated item 'tokio_io::codec::length_delimited::Framed': Moved to tokio-codec
  --> src/server.rs:57:5                                            
   |                                                                        
57 | use tokio_io::codec::length_delimited::Framed;                                              
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                                                                   
   = note: #[warn(deprecated)] on by default                             
                                                                           
warning: use of deprecated item 'tokio_io::codec::length_delimited::Framed': Moved to tokio-codec
   --> src/server.rs:969:37                                                                      
    |                                                           
969 |     inner: WriteBincode<ReadBincode<Framed<I>, Request>, Response>,
    |                                     ^^^^^^^^^                               
                                                                         
warning: use of deprecated item 'tokio_io::codec::length_delimited::Framed': Moved to tokio-codec
   --> src/server.rs:475:55                                                                      
    |                                                            
475 |             inner: WriteBincode::new(ReadBincode::new(Framed::new(socket))),
    |                                                       ^^^^^^^^^^^
                                                                   

```

luser commented:
Thanks for the PR! I think this will need to wait until we land the last few sccache-dist PRs. https://github.com/mozilla/sccache/pull/323 should be the second-to-last one.

joshuawarner32 commented:
Hi! I think I've been running into #248 (which might be fixed by this) - any updates on getting this landed?

luser commented:
We're very close to finishing landing all the distributed compilation work--I'll get this landed as soon as that's all done.

luser commented:
OK, I apologize for how long this has taken, but if you want to rebase this one last time I will merge it immediately.

Marwes commented:
Rebased but seem to have broken something in the distributed compilation. Will try and debug it.

luser commented:
Thanks! Sorry that this took so long to merge that you had to rebase it several times.
